### ISSUE TITLE
several failed units are left after cron commands execution

### ISSUE THREAD

**Issue by wilhelmgoering** [systemd-cron/systemd-cron-next#33]
I'm on an up-to-date Arch linux box (systemd-cron-next version: 1.0.0.rc6-1), with a set of jobs on root's crontab as follows:

```
# min hour dom month dow command
*/5 * * * * script1 >/dev/null 2>&1
*/5 * * * * script2 >/dev/null 2>&1
*/3 * * * * checkupdates | wc -l > /tmp/chkupds 2>&1
* 1 * * *   script3
* 3 * * 0   script4
```

'systemctl --failed' always lists several failed units as follows:

```
● cron-06e13c279b610b7d64c5412ca7e9729b.timer not-found failed failed cron-06e13c279b610b7d64c5412ca7e9729b.ti
● cron-12f71d6fe96b1d8a952b9dea00a2b08a.timer not-found failed failed cron-12f71d6fe96b1d8a952b9dea00a2b08a.ti
● cron-1508990028c238e5e943ea887077cfcc.timer not-found failed failed cron-1508990028c238e5e943ea887077cfcc.ti
● cron-1aa42d0ab12306188535e51c703ff954.timer not-found failed failed cron-1aa42d0ab12306188535e51c703ff954.ti
● cron-1cc8717a11a65cf5cf7f309986f59544.timer not-found failed failed cron-1cc8717a11a65cf5cf7f309986f59544.ti
● cron-23bf0f91c857edf04a4048a568331a85.timer not-found failed failed cron-23bf0f91c857edf04a4048a568331a85.ti
● cron-3877533c7285f861576b4eee0eaa6971.timer not-found failed failed cron-3877533c7285f861576b4eee0eaa6971.ti
● cron-4518e9c1d14310a3d751c2c02365b0ff.timer not-found failed failed cron-4518e9c1d14310a3d751c2c02365b0ff.ti
● cron-72a5dbbed1b97d0cee0ce62c1d21b0e7.timer not-found failed failed cron-72a5dbbed1b97d0cee0ce62c1d21b0e7.ti
● cron-7d5c7ed8286c51ab366bcbb999b453a6.timer not-found failed failed cron-7d5c7ed8286c51ab366bcbb999b453a6.ti
● cron-7e144ed36b5084fcbe3202a9ffbda300.timer not-found failed failed cron-7e144ed36b5084fcbe3202a9ffbda300.ti
● cron-8e377463c5f10f5b0c424a5b5b766d10.timer not-found failed failed cron-8e377463c5f10f5b0c424a5b5b766d10.ti
● cron-a8ce6ec376de5cfaa2b8bfa7b567dcd4.timer not-found failed failed cron-a8ce6ec376de5cfaa2b8bfa7b567dcd4.ti
● cron-abc561fd263fe40ab5bdcce5205ae09b.timer not-found failed failed cron-abc561fd263fe40ab5bdcce5205ae09b.ti
● cron-c253543743bbab935d26f4a27ae14c07.timer not-found failed failed cron-c253543743bbab935d26f4a27ae14c07.ti
● cron-c5dbb45ef40e8cbda29d41a4567ab8e1.timer not-found failed failed cron-c5dbb45ef40e8cbda29d41a4567ab8e1.ti
● cron-ce85dd11cea4c2bce6f698b05cbcd537.timer not-found failed failed cron-ce85dd11cea4c2bce6f698b05cbcd537.ti
● cron-dd7e9019212e9a2b0da1f8d39ec793bb.timer not-found failed failed cron-dd7e9019212e9a2b0da1f8d39ec793bb.ti
● cron-e69e4d1ab4d2651436b41bfdc1299008.timer not-found failed failed cron-e69e4d1ab4d2651436b41bfdc1299008.ti
● cron-e729508d6f7ad9cbe56a72ec9c8aeb7d.timer not-found failed failed cron-e729508d6f7ad9cbe56a72ec9c8aeb7d.ti
```

failed units increase over time.

is this behaviour to be expected?

thanks.

**Comment by kstep** [comment]
No, that's not correct behavior, at least not in the situation as you describe it.

It can happen if you change your crontabs, which leads to timers regenerations. And as systemd remembers all units it ever seen since last boot, timers which correspond to removed crontab lines, will be unavailable after timers regeneration.

Do you change your crontab file somehow? Maybe you generate some crontabs with some automatic process or something?

**Comment by kstep** [comment]
Also I just checked my home server systemd-crontab-generator installation. I have several lines in my root's crontab, and I see three timers which correspond to them. I tried to `systemctl daemon-reload` several times to make timers regenerate, but it all worked fine, not failed timers.

The hex strings you see in unit names are MD5 hashes based on lines in crontab, so timer names can change only if you change your crontab line mirrored by the timer unit.

**Comment by wilhelmgoering** [comment]
I noticed now that failed units appear by just doing 'sudo crontab -e' and
just closing the file without editing it. In fact, having 5 jobs, I get 5
failed units everytime I open the crontab, so 5, 10, 15, and so on.

It's no big deal as long as the system works, but maybe it should re-create
units only after an actual modification to the crontab or just clean up
after itself upon reassessing jobs.



--

"*quia omnino non expedit animabus eorum*"

**Comment by kstep** [comment]
That is strange, the hash shouldn't change if the crontab file is unchanged.
I will look into it. Thanks for the report!
Anyway, it should still work, as long as new active timers are created from crontab file.

**Comment by kstep** [comment]
I guess I have found the origin of the bug.

When you start editing a crontab file, `crontab` makes its temporary copy in `/var/spool/cron`, so `cron-update.path` unit runs a spool dir rescan. Hence, a new timer unit appears for this temporary file. When you exit your editor, this temporary file is removed, which fires `cron-update.path` unit again, so this timer for the temporary file is deleted, but systemd remembers it was available at some point in time.

**Comment by kstep** [comment]
The bug should be fixed in 1.0.0-rc7.

**Comment by wilhelmgoering** [comment]
This is great news. Thank you for the timely fix.


--

"*quia omnino non expedit animabus eorum*"
